src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala

```diff
@@ -0,0 +1,113 @@
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    val sparkSession = spark
+    import sparkSession.implicits._
+    hyperspace = new Hyperspace(spark)
+    FileUtils.delete(new Path(sampleParquetDataLocation))
+    val dfFromSample = sampleData.toDF("Date", "RGUID", "Query", "imprs", "clicks")
+    dfFromSample.write.parquet(sampleParquetDataLocation)
+
+    df = spark.read.parquet(sampleParquetDataLocation)
+    hyperspace.createIndex(df, indexConfig1)
+
+    files = df.queryExecution.optimizedPlan.collect {
+      case LogicalRelation(
+          HadoopFsRelation(
+            location: PartitioningAwareFileIndex,
+            _,
+            dataSchema,
+            _,
+            fileFormat,
+            options),
+          _,
+          _,
+          _) =>
+        location.allFiles.map(FileInfo(_))
+    }.flatten
+    val sourcePath = new Path(files.head.name)
+    val destPath = new Path(files.head.name + ".copy")
+    sourcePath.getFileSystem(new Configuration).copyToLocalFile(sourcePath, destPath)
+  }
+
+  before {}
+
+  after {}
+
+  override def afterAll(): Unit = {
+    super.afterAll()
+    spark.sessionState.conf.unsetConf(IndexConstants.INDEX_HYBRID_SCAN_ENABLED)
+    spark.sessionState.conf.unsetConf("spark.sql.autoBroadcastJoinThreshold")
+    FileUtils.delete(new Path(sampleParquetDataLocation))
+  }
+
+  test("HybridScan filter rule test") {
+    spark.conf.set("spark.hyperspace.index.hybridscan.enabled", false)
+
+    df = spark.read.parquet(sampleParquetDataLocation)
+    val query = df.filter(df("clicks") <= 2000).select(df("query"))
+    val transformed = FilterIndexRule(query.queryExecution.optimizedPlan)
+    assert(transformed.equals(query.queryExecution.optimizedPlan), "No plan transformation.")
+
+    spark.conf.set("spark.hyperspace.index.hybridscan.enabled", true)
+    val planWithHybridScan = FilterIndexRule(query.queryExecution.optimizedPlan)
+    assert(
+      !planWithHybridScan.equals(query.queryExecution.optimizedPlan),
+      "Plan should be transformed.")
+  }
+
+  test("HybridScan join rule test") {
+    spark.conf.set("spark.hyperspace.index.hybridscan.enabled", false)
+    spark.conf.set("spark.sql.autoBroadcastJoinThreshold", -1)
+
+    df = spark.read.parquet(sampleParquetDataLocation)
+    val query = df.filter(df("clicks") <= 2000).select(df("clicks"), df("query"))
+    val query2 = df.filter(df("clicks") >= 4000).select(df("clicks"), df("query"))
+    val join = query.join(query2, "clicks")
+    val transformed = JoinIndexRule(join.queryExecution.optimizedPlan)
+    assert(transformed.equals(join.queryExecution.optimizedPlan), "No plan transformation.")
+
+    spark.conf.set("spark.hyperspace.index.hybridscan.enabled", true)
+    val planWithHybridScan = JoinIndexRule(join.queryExecution.optimizedPlan)
+    assert(
+      !planWithHybridScan.equals(join.queryExecution.optimizedPlan),
+      "Plan should be transformed.")

```

Comment:
Can we validate the changed plan more thoroughly? You can check few examples from [PlannerSuite](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala) or [BucketedReadSuite](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala). I would at least check whether `BucketUnionExec` is introduced along with `Exchange`, and so on.

Also, I think we need to check whether the output is the same when hybrid scan is applied. Can we utilize [`checkAnswer`](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala#L133)?

Reply:
Btw, to compare rows, it may be easier to set up something similar to the existing E2E tests instead of running individual rules.

Reply:
Added some check logic to check the results & physical plan with hybrid scan.